### PR TITLE
[Feature] Support to visualize results based on prediction quality

### PR DESCRIPTION
**Pull request by hhaAndroid**
New add results quality evaluation

### CODE REVIEW

File: tools/analyze_results.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        bad_dir = osp.abspath(osp.join(self.show_dir, 'bad'))
+        self._save_images(good_mAPs, good_dir)
+        self._save_images(bad_mAPs, bad_dir)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description='MMDet eval image prediction result for each')
+    parser.add_argument('config', help='test config file path')
+    parser.add_argument('pkl_path', help='path where test pkl result')
```
Comment (ZwwWayne) [inline tools/analyze_results.py:119]: pkl_path -> prediction-path

File: tools/analyze_results.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    parser.add_argument(
+        'show_dir', help='directory where painted images will be saved')
```
Comment (ZwwWayne) [inline tools/analyze_results.py:121]: show-dir

File: tools/analyze_results.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+    def _show_image_gts_results(self, index, value, show=False, out_dir=None):
+        gts = self.annotations[index]
+        data_info = self.data_infos[index]
+        result = self.results[index]
+        if self.img_prefix is not None:
+            filename = osp.join(self.img_prefix, data_info['filename'])
+        else:
+            filename = data_info['filename']
+        img = mmcv.imread(filename)
```
Comment (ZwwWayne) [inline tools/analyze_results.py:44]: We should use data_pipeline to load image

File: tools/analyze_results.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        # TODO gt mask support
+        imshow_det_bboxes(
```
Comment (ZwwWayne) [inline tools/analyze_results.py:46]: Should we also support to draw GT mask?

File: tools/analyze_results.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+            _mAPs[i] = mAP
+
+        # descending select topk image
+        _mAPs = list(sorted(_mAPs.items(), key=lambda kv: kv[1]))
+
+        # topk
+        good_mAPs = _mAPs[-self.topk:]
+        bad_mAPs = _mAPs[:self.topk]
+
+        good_dir = osp.abspath(osp.join(self.show_dir, 'good'))
```
Comment (ZwwWayne) [inline tools/analyze_results.py:109]: Handle the case when the dataset is too small. For example, topk=20, but the dataset only have 16 images, at this time the the topk should be changed to half of the dataset's length

File: tools/analyze_results.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+import mmcv
+import numpy as np
+from mmcv import Config, DictAction
+
+from mmdet.core.evaluation import eval_map
+from mmdet.core.visualization.image import imshow_det_bboxes
+from mmdet.datasets import build_dataset
+
+
+class ResultEvaluate(object):
```
Comment (ZwwWayne) [inline tools/analyze_results.py:13]: ResultVisualizer

File: tools/analyze_results.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    def _eval_fn(self, det_result, annotation):
+        iou_thrs = np.linspace(
+            .5, 0.95, int(np.round((0.95 - .5) / .05)) + 1, endpoint=True)
+        mean_aps = []
+        for thr in iou_thrs:
+            mean_ap, _ = eval_map([det_result], [annotation], iou_thr=thr)
+            mean_aps.append(mean_ap)
+        return sum(mean_aps) / len(mean_aps)
+
+    def evaluate(self):
```
Comment (ZwwWayne) [inline tools/analyze_results.py:95]: Can we make the method a method of the dataset? Thus for each dataset, we can call it using `dataset.visualize`

File: tools/analyze_results.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+import mmcv
+import numpy as np
+from mmcv import Config, DictAction
+
+from mmdet.core.evaluation import eval_map
+from mmdet.core.visualization.image import imshow_det_bboxes
+from mmdet.datasets import build_dataset
+
+
+def retrieve_loading_pipeline(pipeline):
```
Comment (ZwwWayne) [inline tools/analyze_results.py:13]: May put this function into mmdet.utils

File: tools/analyze_results.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+              annotation,
+              result,
+              class_names=None,
+              show=True,
+              wait_time=0,
+              out_file=None):
+    imshow_det_bboxes(
+        img,
+        annotation['bboxes'],
+        annotation['labels'],
```
Comment (ZwwWayne) [inline tools/analyze_results.py:60]: Can this function show GT masks?

File: tools/analyze_results.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+                filename = data_info['filename']
+            fname, name = osp.splitext(osp.basename(filename))
+            save_filename = fname + '_' + str(round(mAP, 3)) + name
+            out_file = osp.join(out_dir, save_filename)
+
+            visualize(data_info['img'], data_info['ann_info'],
+                      self.results[index], self.CLASSES, self.show,
+                      self.wait_time, out_file)
+
+    def evaluate(self):
```
Comment (ZwwWayne) [inline tools/analyze_results.py:141]: evaluate_and_show

File: tools/analyze_results.py [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+              result,
+              class_names=None,
+              show=True,
+              show_mask=True,
+              wait_time=0,
+              out_file=None):
+    if show_mask:
+        gt_masks = annotation.get('gt_masks', None)
+        if gt_masks is not None:
+            if isinstance(gt_masks, BitmapMasks):
```
Comment (ZwwWayne) [inline tools/analyze_results.py:28]: We can use the API gt_masks.to_array

File: mmdet/datasets/utils.py [hunk 12]
```diff
@@ -60,3 +60,40 @@ def replace_ImageToTensor(pipelines):
                 'data pipeline in your config file.', UserWarning)
             pipelines[i] = {'type': 'DefaultFormatBundle'}
     return pipelines
+
+
+def retrieve_loading_pipeline(pipeline):
+    """Only keep loading image and annotations related configuration
+        Args:
```
Comment (ZwwWayne) [inline mmdet/datasets/utils.py:67]: An empty line and the indent is not correct

File: tools/analyze_results.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+
+def visualize(img,
+              annotation,
+              result,
+              class_names=None,
+              show=True,
+              show_mask=True,
+              wait_time=0,
+              out_file=None):
+    if show_mask:
```
Comment (ZwwWayne) [inline tools/analyze_results.py:24]: Docstring.

File: tools/analyze_results.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+    dataset = build_dataset(cfg.data.test)
+    outputs = mmcv.load(args.prediction_path)
+    result_visualizer = ResultVisualizer(
+        dataset,
+        outputs,
+        topk=args.topk,
+        show=args.show,
+        wait_time=args.wait_time,
+        show_dir=args.show_dir)
+    result_visualizer.evaluate_and_show()
```
Comment (ZwwWayne) [inline tools/analyze_results.py:238]: It seems that the visualizer is general and can be used many times after creation. If so, we may not make dataset and outputs as an attribute to initialize the ResultVisualizer, but make these two arguments of `evaluate_and_show()`.

File: tools/browse_dataset.py [hunk 15]
```diff
@@ -27,7 +29,7 @@ def parse_args():
     parser.add_argument(
         '--show-interval',
         type=float,
-        default=1,
```
Comment (ZwwWayne) [inline tools/browse_dataset.py:30]: Is there any specific reason to change the default behavior?

File: tools/browse_dataset.py [hunk 16]
```diff
@@ -54,11 +56,19 @@ def main():
         filename = os.path.join(args.output_dir,
                                 Path(item['filename']).name
                                 ) if args.output_dir is not None else None
+        gt_masks = item.get('gt_masks', None)
+        if gt_masks is not None:
+            if isinstance(gt_masks, (BitmapMasks, PolygonMasks)):
+                gt_masks = gt_masks.to_ndarray()
+            else:
+                warnings.warn('Unsupported data type')
```
Comment (ZwwWayne) [inline tools/browse_dataset.py:64]: Can we also support nparray or Tensor of gt_masks?

File: tools/analyze_results.py [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+              show_mask=True,
+              wait_time=0,
+              out_file=None):
+    if show_mask:
+        gt_masks = annotation.get('gt_masks', None)
+        if gt_masks is not None:
+            if isinstance(gt_masks, (BitmapMasks, PolygonMasks)):
+                gt_masks = gt_masks.to_ndarray()
+            else:
+                warnings.warn('Unsupported data type')
```
Comment (ZwwWayne) [inline tools/analyze_results.py:30]: Support ndarray and Tensor.

File: tools/browse_dataset.py (hunk 15 above)
Comment (hhaAndroid) [reply tools/browse_dataset.py:30]: Because the window closes too fast,  the display effect is not good

File: mmdet/core/visualization/image.py [hunk 18]
```diff
@@ -169,3 +173,118 @@ def imshow_det_bboxes(img,
...
+        np.full(bbox.shape[0], i, dtype=np.int32)
+        for i, bbox in enumerate(bbox_result)
+    ]
+    labels = np.concatenate(labels)
+    segms = None
+    if segm_result is not None and len(labels) > 0:  # non empty
+        segms = mmcv.concat_list(segm_result)
+        segms = mask_util.decode(segms)
+        segms = segms.transpose(2, 0, 1)
+    imshow_det_bboxes(
```
Comment (ZwwWayne) [inline mmdet/core/visualization/image.py:274]: This function has return, so the return of imshow_det_bboxes should also be returned by this function.

Comment (ZwwWayne) [inline mmdet/core/visualization/image.py:274]: Add an empty line above imshow_det_bboxes

File: tools/analyze_results.py [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+import mmcv
+import numpy as np
+from mmcv import Config, DictAction
+
+from mmdet.core.evaluation import eval_map
+from mmdet.core.visualization import imshow_gt_det_bboxes
+from mmdet.datasets import build_dataset, retrieve_loading_pipeline
+
+
+def _bboxes_map_eval_fn(det_result, annotation):
```
Comment (ZwwWayne) [inline tools/analyze_results.py:13]: _bboxes_map_eval_fn ->  _bboxes_map_eval

File: mmdet/datasets/utils.py [hunk 20]
```diff
@@ -60,3 +60,40 @@ def replace_ImageToTensor(pipelines):
                 'data pipeline in your config file.', UserWarning)
             pipelines[i] = {'type': 'DefaultFormatBundle'}
     return pipelines
+
+
+def retrieve_loading_pipeline(pipeline):
```
Comment (ZwwWayne) [inline mmdet/datasets/utils.py:65]: retrieve_loading_pipeline -> get_loading_pipeline

File: mmdet/datasets/utils.py [hunk 21]
```diff
@@ -60,3 +60,40 @@ def replace_ImageToTensor(pipelines):
...
+        ...    ]
+        >>> assert expected_pipelines ==\
+        ...        retrieve_loading_pipeline(pipelines)
+    """
+    loading_pipeline_cfg = []
+    for cfg in pipeline:
+        if cfg['type'].startswith('Load'):
+            loading_pipeline_cfg.append(cfg)
+    assert len(loading_pipeline_cfg) == 2, \
+        'pipeline cfg must include Loading image and annotations related type'
```
Comment (ZwwWayne) [inline mmdet/datasets/utils.py:98]: this sentence has grammar issues.

File: mmdet/core/visualization/image.py [hunk 22]
```diff
@@ -169,3 +173,118 @@ def imshow_det_bboxes(img,
...
+      font_size (int): Font size of texts. Default: 13
+      win_name (str): The window name. Default: ''
+      fig_size (tuple): Figure size of the pyplot figure. Default: (15, 10)
+      show (bool): Whether to show the image. Default: True
+      wait_time (float): Value of waitKey param. Default: 0.
+      out_file (str or None): The filename to write the image. Default: None
+    """
+    assert 'gt_bboxes' in annotation
+    assert 'gt_labels' in annotation
+    assert isinstance(result, (tuple, list))
```
Comment (ZwwWayne) [inline mmdet/core/visualization/image.py:230]: Add error messages, e.g., what type of result you get.

File: tools/analyze_results.py [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+                          dataset,
+                          results,
+                          topk=20,
+                          show_dir='work_dir',
+                          eval_fn=None):
+        """Evaluate and show results.
+
+        Args:
+            dataset (Dataset): A PyTorch dataset.
+            results (pickle object): pickle object from test results pkl file
```
Comment (ZwwWayne) [inline tools/analyze_results.py:100]: pickle object is not a valid type. It is the original type, for example, a list of tuple or some thing.

File: tools/analyze_results.py [hunk 24]
```diff
@@ -0,0 +1,203 @@
...
+    """Evaluate mAP of single image det result.
+
+    Args:
+        det_result (list[list]): [[cls1_det, cls2_det, ...], ...].
+            The outer list indicates images, and the inner list indicates
+            per-class detected bboxes.
+        annotation (dict): Ground truth annotations where keys of
+             annotations are:
+
+            - `bboxes`: numpy array of shape (n, 4)
```
Comment (ZwwWayne) [inline tools/analyze_results.py:23]: No need to use \` here.

File: tools/browse_dataset.py [hunk 25]
```diff
@@ -54,11 +56,19 @@ def main():
         filename = os.path.join(args.output_dir,
                                 Path(item['filename']).name
                                 ) if args.output_dir is not None else None
+        gt_masks = item.get('gt_masks', None)
```
Comment (ZwwWayne) [inline tools/browse_dataset.py:59]: Use the function `mask2ndarray`.

File: mmdet/core/utils/misc.py [hunk 26]
```diff
@@ -37,3 +40,22 @@ def unmap(data, count, inds, fill=0):
         ret = data.new_full(new_size, fill)
         ret[inds.type(torch.bool), :] = data
     return ret
+
+
+def mask2ndarray(mask):
+    """Mask to ndarray.
+
+    Args:
+        mask (BitmapMasks or PolygonMasks or torch.Tensor or np.ndarray):
```
Comment (ZwwWayne) [inline mmdet/core/utils/misc.py:49]: :obj:\`BitmapMasks\` or :obj:\`PolygonMasks\`

File: mmdet/core/utils/misc.py [hunk 27]
```diff
@@ -37,3 +40,22 @@ def unmap(data, count, inds, fill=0):
...
+         The mask to be converted.
+
+    Returns:
+        np.ndarray: Ndarray mask that has been converted
```
Comment (ZwwWayne) [inline mmdet/core/utils/misc.py:53]: May indicate the shape information

File: mmdet/core/visualization/image.py [hunk 28]
```diff
@@ -169,3 +172,123 @@ def imshow_det_bboxes(img,
...
+          'gt_bboxes' and 'gt_labels' or 'gt_masks'
+      result (tuple[list] or list): The detection result, can be either
+          (bbox, segm) or just bbox.
+      class_names (list[str]): Names of each classes.
+      score_thr (float): Minimum score of bboxes to be shown.  Default: 0
+      gt_bbox_color (str or tuple(int) or :obj:`Color`):Color of bbox lines.
+           The tuple of color should be in BGR order. Default: (255, 102, 61)
+      gt_text_color (str or tuple(int) or :obj:`Color`):Color of texts.
+           The tuple of color should be in BGR order. Default: (255, 102, 61)
+      gt_mask_color (None or str or tuple(int) or :obj:`Color`):
```
Comment (ZwwWayne) [inline mmdet/core/visualization/image.py:209]: (str or tuple(int) or :obj:\`Color\`, optional).

File: mmdet/core/visualization/image.py [hunk 29]
```diff
@@ -64,7 +67,8 @@ def imshow_det_bboxes(img,
         win_name (str): The window name. Default: ''
         fig_size (tuple): Figure size of the pyplot figure. Default: (15, 10)
         wait_time (float): Value of waitKey param. Default: 0.
-        out_file (str or None): The filename to write the image. Default: None
+        out_file (str or optional): The filename to write the image.
```
Comment (ZwwWayne) [inline mmdet/core/visualization/image.py:70]: str, optional

File: mmdet/core/utils/misc.py [hunk 30]
```diff
@@ -37,3 +40,22 @@ def unmap(data, count, inds, fill=0):
         ret = data.new_full(new_size, fill)
         ret[inds.type(torch.bool), :] = data
     return ret
+
+
+def mask2ndarray(mask):
+    """Mask to ndarray.
```
Comment (ZwwWayne) [inline mmdet/core/utils/misc.py:46]: Convert Mask to ndarray.

File: tools/analyze_results.py [hunk 31]
```diff
@@ -0,0 +1,203 @@
...
+import mmcv
+import numpy as np
+from mmcv import Config, DictAction
+
+from mmdet.core.evaluation import eval_map
+from mmdet.core.visualization import imshow_gt_det_bboxes
+from mmdet.datasets import build_dataset, get_loading_pipeline
+
+
+def _bboxes_map_eval(det_result, annotation):
```
Comment (ZwwWayne) [inline tools/analyze_results.py:13]: _bboxes_map_eval -> bbox_map_eval.

File: tests/test_visualization.py [hunk 32]
```diff
@@ -54,8 +55,44 @@ def test_imshow_det_bboxes():
...
+
     # test tensor mask type error
     with pytest.raises(AttributeError):
         segms = torch.tensor(segms)
-        vis.imshow_det_bboxes(
-            image, bbox, label, segms, out_file=tmp_filename, show=False)
+        vis.imshow_det_bboxes(image, bbox, label, segms, show=False)
+
+
+def test_imshow_gt_det_bboxes():
```
Comment (ZwwWayne) [inline tests/test_visualization.py:66]: May also add a single test for mask2ndarray.

**Review by ZwwWayne — APPROVED** [review]
